src/Gaffer/RandomChoice.cpp

```diff
@@ -0,0 +1,355 @@
+{
+	ComputeNode::affects( input, outputs );
+
+	if(
+		input == seedPlug() ||
+		input == seedVariablePlug() ||
+		input == choicesWeightsPlug() ||
+		input == choicesValuesPlug()
+	)
+	{
+		if( const Plug *p = outPlug() )
+		{
+			if( p->children().size() )
+			{
+				for( const auto &cp : Plug::Range( *p ) )
+				{
+					outputs.push_back( cp.get() );
+				}
+			}
+			else
+			{
+				outputs.push_back( p );
+			}
+		}
+	}
+}
+
+void RandomChoice::hash( const ValuePlug *output, const Context *context, IECore::MurmurHash &h ) const
+{
+	ComputeNode::hash( output, context, h );
+
+	if( output == outPlug() || output->parent() == outPlug() )
+	{
+		seedPlug()->hash( h );
+		const string seedVariable = seedVariablePlug()->getValue();
+		if( !seedVariable.empty() )
+		{
+			h.append( context->variableHash( seedVariable ) );
+		}
+		choicesWeightsPlug()->hash( h );
+		choicesValuesPlug()->hash( h );
+		h.append( output->defaultHash() );
+	}
+}
+
+void RandomChoice::compute( ValuePlug *output, const Context *context ) const
+{
+	if( output == outPlug() || output->parent() == outPlug() )
+	{
+		ConstFloatVectorDataPtr weightsData = choicesWeightsPlug()->getValue();
+		const vector<float> &weights = weightsData->readable();
+
+		size_t seed = seedPlug()->getValue();
+		const string seedVariable = seedVariablePlug()->getValue();
+		if( !seedVariable.empty() )
+		{
+			/// \todo See comments in `Random::computeSeed()`.
+			IECore::DataPtr contextData = context->getAsData( seedVariable, nullptr );
+			if( contextData )
+			{
+				const MurmurHash hash = contextData->Object::hash();
+				seed += hash.h1();
+			}
+		}
+		Imath::Rand48 random( seed );
+
+		dispatchPlugFunction(
+			outPlug(),
+			[this, output, &weights, &random] ( auto plug ) {
+
+				using PlugType = remove_const_t<remove_pointer_t<decltype( plug )>>;
+				using ValueType = typename PlugType::ValueType;
+
+				auto choicesData = this->choicesValuesPlug<ValuesPlugType<ValueType>>()->getValue();
+				const auto &choices = choicesData->readable();
+				if( !choices.size() )
+				{
+					output->setToDefault();
+					return;
+				}
+
+				// We might have been provided with too few weights or too many. We
+				// try to be helpful by using defaults for missing weights, and ignoring
+				// any extras.
+				const float defaultWeight = 1.0f;
+				float weightsSum = accumulate(
+					weights.begin(), weights.begin() + min( weights.size(), choices.size() ), 0.0f
+				);
+				if( weights.size() < choices.size() )
+				{
+					weightsSum += defaultWeight * ( choices.size() - weights.size() );
+				}
+
+				if( weightsSum == 0.0f )
+				{
+					output->setToDefault();
+					return;
+				}
+
+				const float r = random.nextf( 0, weightsSum );
+				float s = 0;
+				for( size_t i = 0; i < choices.size(); ++i )
+				{
+					s += weights.size() > i ? weights[i] : defaultWeight;
+					if( s >= r )
+					{
+						setValue( plug, choices[i], output );
+						return;
+					}
+				}
```

Comment:
If this gets used in really heavy instancing setups, it could be fairly valuable to have a cache for a normalized accumulated weights vector, so you could use an binary/interpolation search rather than stepping through everything, but there's nothing that prevents us doing that later.

Reply:
I did actually try that, but it only showed any benefit for very big lists of choices. I seem to have lost my notes, but my recollection was that I needed `choices.values` to be in the order of 100,000 to see a 2x speedup. I found it hard to imagine more than a few hundred or maybe a thousand distinct values, so went with the simpler code for now.

Reply:
If you went to trouble of testing that, would be nice to have it noted in the code, so the next person to come along doesn't wonder if it's a bottleneck.